src/validator.js

```diff
@@ -47,6 +50,92 @@ export default class Validator {
+      .then((metadata) => {
+        if (metadata.hasOwnProperty(installRdfPath)) {
+          return this.xpi.getFileAsString(installRdfPath)
+            .then((content) => {
+              var rdfScanner = new RDFScanner(content, installRdfPath);
+              return rdfScanner.getXMLDoc();
+            })
+            .then((xmlDoc) => {
+              // Lookup addon type from install.rdf.
+              return new Promise((resolve) => {
+                var typeNodes = xmlDoc.getElementsByTagName('em:type');
+                if (typeNodes.length > 1) {
+                  throw new Error('Multiple <em:type> elements found');
+                }
+                var node = typeNodes[0];
+                if (node && node.firstChild && node.firstChild.nodeValue) {
+                  var typeValue = node.firstChild.nodeValue;
+                  var resolvedTypeValue = null;
+                  if (!constants.INSTALL_RDF_TYPE_MAP
+                        .hasOwnProperty(typeValue)) {
+                    log.debug('Invalid type value "%s"', typeValue);
+                    this.collector.addError(messages.TYPE_INVALID);
+                  } else {
+                    var resolvedTypeValue =
+                      constants.INSTALL_RDF_TYPE_MAP[typeValue];
+                    log.debug('Mapping original <em:type> value "%s" -> "%s"',
+                              typeValue, resolvedTypeValue);
+                  }
+                  // This will resolve with null or the actual value.
+                  resolve(resolvedTypeValue);
+                } else {
+                  log.warn('<em:type> was not found in install.rdf');
+                  this.collector.addNotice(messages.TYPE_MISSING);
+                  resolve(null);
+                }
+              });
+            });
+        } else {
+          log.warn('No install.rdf was found in the package metadata');
+          this.collector.addNotice(messages.TYPE_NO_INSTALL_RDF);
+          return Promise.resolve(null);
+        }
+      }).then((packageType) => {
+        if (packageType === null) {
+          log.info(singleLineString`Determining addon type from install.rdf
+                   failed. Guessing from layout`);
+          return this.xpi.getMetaData()
+            .then((metadata) => {
+              for (let path_ of Object.keys(metadata)) {
+                if (path_.startsWith('dictionaries/')) {
+                  return Promise.resolve(
+                    constants.PACKAGE_DICTIONARY);
+                }
+              }
+
+              // TODO: This follows the amo-validator approach.
+              // There's no type element, so the spec says that it's either a
+              // theme or an extension. At this point, we know that it isn't
+              // a dictionary, language pack, or multiple extension pack.
+              log.warn('Falling back to decide type from package extension');
+
+              var extensions = {
+                '.jar': '4',
```

Comment:
Yeah good idea. Copied this pretty much verbatim from the python code but that would save some look-ups.